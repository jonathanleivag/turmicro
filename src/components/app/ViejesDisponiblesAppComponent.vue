<script setup lang="ts">
import { defineProps, onMounted, ref } from 'vue'
import Web3 from 'web3/dist/web3.min.js'

import { Turmicro } from '../../interfaces/ITurmicro'
import { useStore } from '../../store'
import { weiToEther } from '../../utils'
import CardUiComponent from '../ui/CardUiComponent.vue'

interface IProps {
  deployContract: Turmicro
  web3: Web3
  account: string
}

interface IPropsViajes {
  name: string
  price: number
  puntos: number
}

const props = defineProps<IProps>()
const viajes = ref<IPropsViajes[]>([])
const store = useStore()

onMounted(() => {
  getViaje()
})

const getTotalViajes = async (): Promise<number> =>
  +(await props.deployContract.getTotalViajes())

const getViaje = async () => {
  const totalViajes = await getTotalViajes()
  for (let i = 0; i < totalViajes; i++) {
    const viaje = await props.deployContract._viajes(i)
    viajes.value.push({
      name: viaje.nombre,
      price: +viaje.valor,
      puntos: +viaje.puntos
    })
  }
}

const comprar = async (
  indexViaje: number,
  value: number,
  name: string,
  puntos: number
) => {
  await props.deployContract.pagarViaje(indexViaje, {
    from: props.account,
    value
  })

  store.dispatch('tusViajes/setOneViaje', {
    nombre: name,
    valor: value,
    puntos
  })
}
</script>

<template>
  <CardUiComponent title="Viajes disponibles">
    <ul class="card__list">
      <li v-for="(viaje, index) in viajes" :key="viaje.name">
        <span>Nombre: {{ viaje.name }}</span>
        <span>Precio: {{ weiToEther(web3, viaje.price) }} ETH</span>
        <span>Puntos: {{ viaje.puntos }}</span>
        <div class="card__mantenedor_boton">
          <button
            @click="comprar(index, viaje.price, viaje.name, viaje.puntos)"
            typer="button"
            class="card__comprar"
          >
            Comprar
          </button>
        </div>
      </li>
    </ul>
  </CardUiComponent>
</template>
