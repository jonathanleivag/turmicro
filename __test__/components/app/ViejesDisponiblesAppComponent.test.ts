// ViejesDisponiblesAppComponent

import { mount } from '@vue/test-utils'

import ViejesDisponiblesAppComponent from '../../../src/components/app/ViejesDisponiblesAppComponent.vue'

describe('ViejesDisponiblesAppComponent', () => {
  const wrapper = mount(ViejesDisponiblesAppComponent)

  test('Snapshot de ViejesDisponiblesAppComponent', () => {
    expect(ViejesDisponiblesAppComponent).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('El titulo tiene que ser Puntos acomulados', () => {
    expect(wrapper.find('h3').text()).toBe('Viajes disponibles')
  })
})
