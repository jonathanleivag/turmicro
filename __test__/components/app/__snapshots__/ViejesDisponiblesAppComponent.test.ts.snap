// Vitest Snapshot v1

exports[`ViejesDisponiblesAppComponent > Snapshot de ViejesDisponiblesAppComponent 1`] = `
"<div class=\\"card\\">
  <div class=\\"card__header\\">
    <h3 class=\\"card__title\\">Viajes disponibles</h3>
  </div>
  <div class=\\"card__body\\">
    <ul class=\\"card__list\\"></ul>
  </div>
</div>"
`;
